stac_geoparquet/to_arrow.py

```diff
@@ -0,0 +1,286 @@
+    path: Union[str, Path],
+    *,
+    chunk_size: int = 8192,
+    schema: Optional[pa.Schema] = None,
+):
+    # Define outside of if/else to make mypy happy
+    items: List[dict] = []
+
+    # If the schema was not provided, then we need to load all data into memory at once
+    # to perform schema resolution.
+    if schema is None:
+        with open(path) as f:
+            for line in f:
+                items.append(json.loads(line))
+
+        return parse_stac_items_to_arrow(items, chunk_size=chunk_size, schema=schema)
+
+    # Otherwise, we can stream over the input, converting each batch of `chunk_size`
+    # into an Arrow RecordBatch at a time. This is much more memory efficient.
+    with open(path) as f:
+        batches: List[pa.RecordBatch] = []
+        for line in f:
+            items.append(json.loads(line))
+
+            if len(items) >= chunk_size:
+                batches.append(_stac_items_to_arrow(items, schema=schema))
+                items = []
+
+    # Don't forget the last chunk in case the total number of items is not a multiple of
+    # chunk_size.
+    if len(items) > 0:
+        batches.append(_stac_items_to_arrow(items, schema=schema))
+
+    stac_table = pa.Table.from_batches(batches, schema=schema)
+    return _process_arrow_table(stac_table)
+
+
+def _process_arrow_table(table: pa.Table) -> pa.Table:
+    table = _bring_properties_to_top_level(table)
+    table = _convert_timestamp_columns(table)
+    table = _convert_bbox_to_struct(table)
+    return table
+
+
+def _stac_items_to_arrow(
+    items: Sequence[Dict[str, Any]], *, schema: Optional[pa.Schema] = None
+) -> pa.RecordBatch:
+    """Convert dicts representing STAC Items to Arrow
+
+    This converts GeoJSON geometries to WKB before Arrow conversion to allow multiple
+    geometry types.
+
+    All items will be parsed into a single RecordBatch, meaning that each internal array
+    is fully contiguous in memory for the length of `items`.
+
+    Args:
+        items: STAC Items to convert to Arrow
+
+    Kwargs:
+        schema: An optional schema that describes the format of the data. Note that this
+            must represent the geometry column as binary type.
+
+    Returns:
+        Arrow RecordBatch with items in Arrow
+    """
+    # Preprocess GeoJSON to WKB in each STAC item
+    # Otherwise, pyarrow will try to parse coordinates into a native geometry type and
+    # if you have multiple geometry types pyarrow will error with
+    # `ArrowInvalid: cannot mix list and non-list, non-null values`
+    wkb_items = []
+    for item in items:
+        wkb_item = deepcopy(item)
+        # Note: this mutates the existing items. Should we
+        wkb_item["geometry"] = shapely.to_wkb(
+            shapely.geometry.shape(wkb_item["geometry"]), flavor="iso"
+        )
+        wkb_items.append(wkb_item)
+
+    if schema is not None:
+        array = pa.array(items, type=pa.struct(schema))
+    else:
+        array = pa.array(items)
+
```

Comment:
Maybe just need to use `wkb_items` here?
```suggestion
        array = pa.array(wkb_items, type=pa.struct(schema))
    else:
        array = pa.array(wkb_items)

```

Reply:
Oh oops 🤦‍♂️

Reply:
yeah I switched to copying the input instead of modifying the user's existing dicts, and forgot to update it here